src/lib/ImpToLLVM.hs

```diff
@@ -1096,10 +1077,48 @@ pointerType = const L.PointerType
 pointerType = L.PointerType
 #endif
 
+-- === Output stream ===
+
+getOutputStream :: LLVMBuilder m => m Operand
+getOutputStream = getDyvar OutStreamDyvar
+
 initializeOutputStream :: LLVMBuilder m => Operand -> m ()
 initializeOutputStream streamFD = do
   streamPtr <- emitExternCall fdopenFun [streamFD]
-  store outputStreamPtr streamPtr
+  setDyvar OutStreamDyvar streamPtr
+
+-- === Dynamically-scoped vars via thread-local pthreads vars ===
+
+dyvarDefs :: [L.Definition]
+dyvarDefs  = (:[]) $ L.GlobalDefinition $ L.globalVariableDefaults
+  { L.name = dynamicVarLName OutStreamDyvar
+  , L.type' = hostVoidp
+  , L.linkage = L.External
+  , L.initializer = Nothing }
+
+dynamicVarLName :: DynamicVar -> L.Name
+dynamicVarLName v = fromString $ dynamicVarCName v
+
+setDyvar :: LLVMBuilder m => DynamicVar -> Operand -> m ()
+setDyvar v val = do
+  key <- getDyvarKey v
+  void $ emitExternCall pthread_setspecific [key, val]
+
+getDyvar :: LLVMBuilder m => DynamicVar -> m Operand
+getDyvar v = do
+  key <- getDyvarKey v
+  emitExternCall pthread_getspecific [key]
+
+getDyvarKey :: LLVMBuilder m => DynamicVar -> m Operand
+getDyvarKey v = do
+  let ref = globalReference (hostPtrTy hostVoidp) $ dynamicVarLName v
+  load hostVoidp (L.ConstantOperand ref)
+
+pthread_setspecific :: ExternFunSpec
+pthread_setspecific = ExternFunSpec "pthread_setspecific" i32 [] [] [hostVoidp, hostVoidp]
```

Comment:
This code makes me a little nervous. Those pthread functions take the key as a value, and you've declared it as a `void*` here. But it's often not defined as a `void*`! if you look [here](https://elixir.bootlin.com/glibc/latest/source/sysdeps/htl/bits/types/__pthread_key.h#L22) it's actually an `int`!

One reasonable way to handle the vagueness of the pthread standard would be to assume that it's an int here, and then include a `static_assert` in `dexrt` that verifies that it truly is an int. So, in this file:
```hs
-- This is actually implementation-dependent, but we verify that it's true in dexrt.cpp
pthreadKeyTy :: L.Type
pthreadKeyTy = i32
```
and then in `dexrt.cpp`:
```cpp
#include <type_traits>
static_assert(std::is_same<pthread_key_t, int>::value, "Expected pthread_key_t to be an int!");
static_assert(sizeof(int) == 4, "Expected int to be 4B!");
```

Reply:
Great catch. Thanks for figuring that out. I'll do as you suggest.